### ISSUE TITLE
Add to Outlook not working in IE

### ISSUE THREAD

**Issue by commonpike** [commonpike/add-to-calendar-buttons#1]
From https://github.com/carlsednaoui/add-to-calendar-buttons/issues/15

.. but I need more info. What is (not) happening ? How can you repeat that ?

**Comment by LittleDogFido** [comment]
If you indeed have it working in IE and Edge then that is very encouraging. Using the tester link you sent me, when I attempt to download the ,ics file via the Outlook link(or iCal) in IE I am redirected to a "This webpage cannot be displayed" error page with the "most likely cause" stating that some content requires a program I don't have installed. It shouldn't really need a program as far as I know, and just download the .ics as it does with the other browsers, but for what its worth I do have Outlook. The following appears in the address bar of the error page:

> data:text/calendar;charset=utf8,BEGIN:VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0AURL:https://commonpike.github.io/add-to-calendar-buttons/tester.html%0ADTSTART:20190101T000000Z%0ADTEND:20190101T010000Z%0ASUMMARY:OUI2%20Add%20To%20Calender%20Test%2001%0ADESCRIPTION:%0ALOCATION:%0AUID:-https://commonpike.github.io/add-to-calendar-buttons/tester.html%0AEND:VEVENT%0AEND:VCALENDAR


Additionally, in IE the widgets for the events with a location(Toronto) do not display although the headers for them do.  In regards to Edge, an error appears at the bottom of the screen stating that event-invite.ics couldn't be downloaded when Outlook or iCal is selected, although there is no prompt for more details. Prior to updating to the latest version of Edge, nothing happened at all when I clicked the Outlook or iCal link. In all cases the links for the web-based calendars work just fine. I was still unable to download the ics on IE or Edge prior to your overhaul and I had the "nothing happens" issue with Chrome prior to the adding of the download attribute. I am currently using the most recent version of both IE and Edge and have checked the settings of both to make sure there isn't anything abnormal.

**Comment by commonpike** [comment]
OK - thanks for the clear report.

As you found out, OUICal generates data-uri links that, when you click them, download the thing described after the 'data:' part of the string. But it seems support is limited in IE and even edge
https://caniuse.com/#feat=datauri

As you see, edge 75 won't have that problem (of course edge 75 will just be chrome).

But I also see the blob workaround you referred to earlier
https://github.com/jasonsalzman/react-add-to-calendar/issues/4

That should be doable. I will look into getting that in - I'm using this myself - but no promise on when.

**Comment by LittleDogFido** [comment]
Thanks so much for your help and the information. At the very least now I understand why the problem exists, didn't realize there was limited support. 

I look forward to when you are able to get around to getting the workaround in. I'm attempting to add in the calendar functionality as part of an(relatively short) internship I'm doing, so I'll continue working on my own to integrate the work around. Thanks again.

**Comment by commonpike** [comment]
Hm .. ok .. let me sketch what I think I would do. I would want the script to write a `javascript:` url instead of a `data:` url *only* if the data: uri is not supported *and* the blob solution is supported. 

The javascript would be a function that creates a blob and calls msSaveOrOpenBlob.

So in the exports, add a function like 'msDownloadCalendar' or so

like, in line 392,

```
 exports.msDownloadCalendar = function(url) {
    if (thisHasToBeDoneAndCanBeDone) {
     var blob = new Blob([url], { type: 'text/calendar' });
     window.navigator.msSaveOrOpenBlob(blob, 'download.ics');
  }
};
```

Then, in the calendarGenerators.ics .. line 176 .. 
```
 if (thisHasToBeDoneAndCanBeDone) {
    return '<a class="' + eClass + '" href="javascript:msDownloadCalendar(\'' +
        href + '\')">' + calendarName + '</a>';
} else {
   return '<a class="' + eClass + '" download="'+CONFIG.texts.download+'" href="' +
        href + '">' + calendarName + '</a>';
}
```

now, if you load the page, on the ics link, it should say javascript:msDownloadCalendar(...) as a link, and if you click it, it should call the function, which should then exist.

now that is the theory. the real time is going to go into figuring out when thisHasToBeDoneAndCanBeDone , and what happens with single quotes, double quotes, onclick, return values .. 

let me know how you fare !

**Comment by LittleDogFido** [comment]
Greetings,

I have been plugging away at this during the week and have made some
progress, at least in terms of my understanding, but not as much as I had
hoped.

I have replaced (thisHasToBeDoneAndCanBeDone)
with (navigator.userAgent.indexOf("MSIE") != -1) which works for detecting
whether or not the browser is IE. Additionally, from my understanding of
the code and research into somewhat similar problems it seems to me that '<a
class="' + eClass + '" href="javascript:msDownloadCalendar">' +
calendarName + '</a>'
would be sufficient to execute the function. It is also my understanding
that the "url" in var blob = new Blob([url] should be a new string
containing everything from 'BEGIN:VCALENDAR' to 'END:VCALENDAR' but despite
removing encode URI which IE does not recognize, adding ;charset=utf8
after 'text/calendar',
and tweaking things in different ways I seem to be unable to get IE to
create the file, regardless of the other conditions. The frustrating thing
about this is that in all of my poking and prodding I did, at one point,
have things such that it created the file and prompted to downoad when
tester.html was opened but was not embedded in a link. Unfortunately I was
changing things up so much that I don't even remember how I accomplished
this haha.

The real issue that I am running up against, that is preventing from
figuring the proceeding issues out through trial and error is that most
anything I add to the file, completely destroys all functionality as the
console registers either createCalendar and/or addToCalendar as undefined.
Which is effected seems to change depending on what of the
changes/additions are attempted and IE is much more susceptible to this as
sometimes things will still work perfectly fine on Chrome, yet receive the
error on IE.

I'm continuing to go add it, and hopefully learn along the way, but I was
curious if you had any thoughts on why I might be encountering the
createCalendar/adToCalendar undefined issue.

Thanks,
Dave

**Comment by commonpike** [comment]
My problem here is, I don't have the hardware to test it. 
Could  you try this file for me:

https://raw.githubusercontent.com/commonpike/add-to-calendar-buttons/feature/msblob/add-to-calendar.js

Here you can see what I did. 

https://github.com/commonpike/add-to-calendar-buttons/compare/master...commonpike:feature/msblob

Its basicly what I described. _thisHasToBeDoneAndCanBeDone_ is a tedious thing. It can be done only if 'msSaveOrOpenBlob' is supported by your browser. It has to be done only if your browser does not support the proper data-uri. If it has to be done but it can't be done, there now is a javascript warning - that should not happen, be never say never.

> The real issue that I am running up against, that is preventing from
> figuring the proceeding issues out through trial and error is that most
> anything I add to the file, completely destroys all functionality as the
> console registers either createCalendar and/or addToCalendar as undefined.

Welcome to the world of debugging. It depends on where you make the typo - the console should tell you ! - what happens. If for some reason the javascript can't be parsed (like, missing semicolons or surplus whitespace in the middle of lines, nothing will work. 

Can you let me know if my version works ?

(What I expect, is that it actually downloads something, but I'm not 100% sure if its a valid calendar).

**Comment by LittleDogFido** [comment]
You are correct that it downloads something when I click the link, but that
the file does not appear to be a valid calendar. Nothing is added to the
calendar when it is opened in Outlook and Window Calender says "Couldn't
open event. It is possible the calendar file is corrupted".

When I initially opened tester.html in IE it had the same issue as mine,
the console indicated that addToCalender was undefined and the boxes did
not show up. After taking a look at tester.html I noticed that the <div>
tags were nested inside <p> tags which to my understanding doesn't work.
After replacing the <p> tags with <div>s the drop downs displayed, except
for some reason those whose events had a timezone attached. Oddly, it
appears that for IE the <div>  fix only works for tester.html but not
example.html, and I am unsure why the <p> tags were not an issue for
Chrome. As far as the drop downs not displaying in IE events with a
timezone , the console gives an error of:

SCRIPT5118: Option value 'AMERICA/TORONTO' for 'timeZone' is outside of
valid range. Expected: ['UTC']
add-to-calendar.js (217,9)

When it comes to downloading the actual file, when copy the Outlook link
from one of the working dropdowns in IE what I get is:

ieDownloadCalendar('data%3Atext%2Fcalendar%3Bcharset%3Dutf8%2CBEGIN%3AVCALENDAR%250AVERSION%3A2.0%250ABEGIN%3AVEVENT%250AURL%3Afile%3A%2F%2FC%3A%255CUsers%255CCoveyD%255CDesktop%255Cadd-to-calendar-buttons-masterPike%255Cadd-to-calendar-buttons-master%255Ctester.html%250ADTSTART%3A20190101T000000Z%250ADTEND%3A20190101T020000Z%250ASUMMARY%3AOUI2%2520Add%2520To%2520Calender%2520Test%252003%250ADESCRIPTION%3A%250ALOCATION%3A%250AUID%3A-file%3A%2F%2FC%3A%255CUsers%255CCoveyD%255CDesktop%255Cadd-to-calendar-buttons-masterPike%255Cadd-to-calendar-buttons-master%255Ctester.html%250AEND%3AVEVENT%250AEND%3AVCALENDAR')

Whereas on Chrome the link reads:

data:text/calendar;charset=utf8,BEGIN:VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0AURL:file:///C:/Users/CoveyD/Desktop/add-to-calendar-buttons-masterPike/add-to-calendar-buttons-master/tester.html%0ADTSTART:20190101T000000Z%0ADTEND:20190101T020000Z%0ASUMMARY:OUI2%20Add%20To%20Calender%20Test%2003%0ADESCRIPTION:%0ALOCATION:%0AUID:-file:///C:/Users/CoveyD/Desktop/add-to-calendar-buttons-masterPike/add-to-calendar-buttons-master/tester.html%0AEND:VEVENT%0AEND:VCALENDAR


Should the links essentially be the same except for data: being replaced
with Javascript function? or should they be very different as they are? If
they should be similar, would changing encodeURIcomponent to
decodeURIcomponent fix that?

I don't know if this is useful information but also, when I run the Chrome
link through urldecoder.org I get:

data:text/calendar;charset=utf8,BEGIN:VCALENDAR
VERSION:2.0
BEGIN:VEVENT
URL:file:///C:/Users/CoveyD/Desktop/add-to-calendar-buttons-masterPike/add-to-calendar-buttons-master/tester.html
DTSTART:20190101T000000Z
DTEND:20190101T020000Z
SUMMARY:OUI2 Add To Calender Test 03
DESCRIPTION:
LOCATION:
UID:-file:///C:/Users/CoveyD/Desktop/add-to-calendar-buttons-masterPike/add-to-calendar-buttons-master/tester.html
END:VEVENT
END:VCALENDAR

When I do the same to the IE link, removing  "ieDownloadCalender" and well
as the single quotes and parentheses at either end I get something that
looks like the original Chrome link. When I think run THAT result  through
the decoder I get the following:

data:text/calendar;charset=utf8,BEGIN:VCALENDAR
VERSION:2.0
BEGIN:VEVENT
URL:file://C:\Users\CoveyD\Desktop\add-to-calendar-buttons-masterPike\add-to-calendar-buttons-master\tester.html
DTSTART:20190101T000000Z
DTEND:20190101T020000Z
SUMMARY:OUI2 Add To Calender Test 03
DESCRIPTION:
LOCATION:
UID:-file://C:\Users\CoveyD\Desktop\add-to-calendar-buttons-masterPike\add-to-calendar-buttons-master\tester.html
END:VEVENT
END:VCALENDAR

Now at first these two looked identical to me, but after closer inspection
I realized that on the decoded Chrome link there are three forward slashes
" / " following UID-file and URL-file whereas the IE link has only two.
Might this be part of the reason why the ICS file does not work, despite
apparently containing the correct data for the file?


Hope that this is all helpful information. I will keep poking and prodding
away.

**Comment by commonpike** [comment]
Bummer

> Should the links essentially be the same except for data: being replaced
> with Javascript function? or should they be very different as they are? 

That is on purpose. If you check what the function ieDownloadCalendar 
was receiving (by eg inserting ``console.log(url);``) you should see it is
the same as the chrome url (except for the forward slashes, which is 
a browser difference).

> If
> they should be similar, would changing encodeURIcomponent to
> decodeURIcomponent fix that?

No - removing encodeURIcomponent might help on IE, but it breaks the link on other browsers. 

I've now changed 'href' to 'onclick' so I can drop the extra encodeURIcomponent. Can you try with the latest version here 

https://raw.githubusercontent.com/commonpike/add-to-calendar-buttons/feature/msblob/add-to-calendar.js

and see if that helped ? You can see the changes here

https://github.com/commonpike/add-to-calendar-buttons/commit/49ea867a48793bfede4f420a16a84cad2a44c5a9

**Comment by commonpike** [comment]
As for the other errors, I've added issues here https://github.com/commonpike/add-to-calendar-buttons/issues

**Comment by commonpike** [comment]
And if the new version doesn't help. can you open the console and just type in 

```
ieDownloadCalendar('foo');
```
(that will give a nice error)
and then try

```
ieDownloadCalendar('data:text/calendar;charset=utf8,BEGIN:VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0ADTSTART:20190101T000000Z%0ADTEND:20190101T020000Z%0ASUMMARY:OUI2%20Add%20To%20Calender%20Test%2003%0ADESCRIPTION:%0ALOCATION:%0AUID:9999%0AEND:VEVENT%0AEND:VCALENDAR');
```

**Comment by LittleDogFido** [comment]
Unfortunately the new version does not help and the .ics file that
downloads doesn't work. Entering  ieDownloadCalendar('foo'); just yields
undefined and attempts to download a .ics file. Entering
ieDownloadCalendar('data:text/calendar;charset=utf8,BEGIN:
VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0ADTSTART:20190101T000000Z%0ADTEND:
20190101T020000Z%0ASUMMARY:OUI2%20Add%20To%20Calender%
20Test%2003%0ADESCRIPTION:%0ALOCATION:%0AUID:9999%0AEND:VEVENT%0AEND:VCALENDAR');
again gives a readout of undefined and prompts for the download of an ics
file but it too does not work. Not that I was expecting different results
but tried the same things in the console using your previous version,
without the onclick, and got the same results.

Something interesting I've noticed about file size that I don't think would
be relevant to troubleshooting but figured I'd mention anyways. When
downloading on Chrome the ics file it creates is 424 bytes. Before your
"onclick" edit the file created in IE was 488 bytes and after the edit it
was 510 bytes. When entering ieDownloadCalendar('foo'); the file it tries
to create is 3 bytes and when entering ieDownloadCalendar with the longer
string of information the file size of the ics it tries to create is 244
bytes.

**Comment by commonpike** [comment]
> Entering  ieDownloadCalendar('foo'); just yields
> undefined and attempts to download a .ics file. 

Of 3 bytes: the bytes 'f','o' and 'o' :-)

> Entering
> ieDownloadCalendar('...');
> again gives a readout of undefined and prompts for the download of an ics
> file but it too does not work. 

I was hoping Outlook would give a clear error message ? Perhaps not.
From this point on, we are debugging what Outlook expects to be in the ics file, not how the webpage works.

Can you try 
```
ieDownloadCalendar('BEGIN:VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0ADTSTART:20190101T000000Z%0ADTEND:20190101T020000Z%0ASUMMARY:OUI2%20Add%20To%20Calender%20Test%2003%0ADESCRIPTION:%0ALOCATION:%0AUID:9999%0AEND:VEVENT%0AEND:VCALENDAR');
```
instead ?

**Comment by LittleDogFido** [comment]
That too was unfortunately ineffective. Going off what you said about what
Outlook expects to be in the ICS file, I downloaded the Outlook ICS
generated from Chrome, and the one generated by IE and opened them both in
a text editor to compare.

The one from Chrome had everything on separate lines and read:

BEGIN:VCALENDAR
VERSION:2.0
BEGIN:VEVENT
URL:file:///C:/Users/CoveyD/Desktop/add-to-calendar-buttons-masterPike%20-%20Copy/add-to-calendar-buttons-master/tester.html
DTSTART:20190101T000000Z
DTEND:20190101T010000Z
SUMMARY:OUI2 Add To Calender Test 01
DESCRIPTION:
LOCATION:
UID:-file:///C:/Users/CoveyD/Desktop/add-to-calendar-buttons-masterPike%20-%20Copy/add-to-calendar-buttons-master/tester.html
END:VEVENT
END:VCALENDAR

So, the name thing as when I ran the URL through urldecoder.org

Whereas the ics generated in IE reads this when opened in the text editor:

data:text/calendar;charset=utf8,BEGIN:VCALENDAR%0AVERSION:2.0%0ABEGIN:VEVENT%0AURL:file://C:%5CUsers%5CCoveyD%5CDesktop%5Cadd-to-calendar-buttons-masterPike%20-%20Copy%5Cadd-to-calendar-buttons-master%5Ctester.html%0ADTSTART:20190101T000000Z%0ADTEND:20190101T010000Z%0ASUMMARY:OUI2%20Add%20To%20Calender%20Test%2001%0ADESCRIPTION:%0ALOCATION:%0AUID:-file://C:%5CUsers%5CCoveyD%5CDesktop%5Cadd-to-calendar-buttons-masterPike%20-%20Copy%5Cadd-to-calendar-buttons-master%5Ctester.html%0AEND:VEVENT%0AEND:VCALENDAR

I did a bit of fiddling around with it. Just removing
"data:text/calendar;charset=utf8," had no effect in Outlooks ability to
recognize the file. When I remove it AND separate out the lines, with each
one still preceded by %0A IE will at least recognize that I'm trying to
open an ICS but but read "The Internet Calendar file IE.ics does not
contain any appointments, do you want to add it to your calendar?".
Removing the "%0A" from each line allows Outook to recognize the date and
time of the calendar, though the spaces still display as "%20" in the
subject.

These %_ are all part of the URL encoding correct? It would seem that using
decodeURIcomponent would resolve that, but as you said would break it it
other browsers. It would also not resolve the issue of each item needing to
be on a new line. Is .replace() capable of making these changes?

**Comment by commonpike** [comment]
Aha, and thanks for testing. I was assuming url encoding was needed because that is what they do here https://github.com/jasonsalzman/react-add-to-calendar/issues/4 - however, i see that that message ends in 'help someone' :-) 

Could you try if this downloads a valid calendar event:

```
ieDownloadCalendar('BEGIN:VCALENDAR\nVERSION:2.0\nBEGIN:VEVENT\nDTSTART:20190101T000000Z\nDTEND:20190101T010000Z\nSUMMARY:OUI2 Add To Calender Test 01\nDESCRIPTION:\nLOCATION:\nUID:99999\nEND:VEVENT\nEND:VCALENDAR')
```

(here, the text is not urlencoded, and newlines are written as `\n`, which is correct for javascript).
If that works, it's easy to rewrite the code (but not using urldecode, because doh we just urlencoded it ourselves)

**Comment by LittleDogFido** [comment]
What you sent does indeed download a valid calendar entry! It opens at
Outlook and I am able to add it to the calendar just fine. Would it still
need to be encoded in the Outlook download for Chrome?

**Comment by commonpike** [comment]
OK - can you try this one:

https://github.com/commonpike/add-to-calendar-buttons/raw/feature/msblob/add-to-calendar.js

You can see what I did here:

https://github.com/commonpike/add-to-calendar-buttons/commit/2fb0742b212a028a2cf300ec3437f485af77122f

> Would it still
> need to be encoded in the Outlook download for Chrome?

No, apparently Outlook expects a 'vanilla' textfile. This is regardless of who downloads that file (outlook doesnt know where it comes from). In the code, it needs to be escaped (instead of encoded), to be able to pass through javascript on one hand and html attributes on the other. I hope that all works now.

**Comment by LittleDogFido** [comment]
Great New! It all seems be working now. I've tested it on Chrome, Firefox,
IE, Edge, and Safari on PC, as well as Safari and Chrome on Mac. Mobile
compatibility seems to be a different story, though I'm not sure how much
of a priority that is. I had tested the ICS file previously, in the
original version created by carlsednaoui and it worked fine. Comparing the
two generated files in a mobile text editor the issue seems to be the
"UID:" line in the ICS file as it was not present in the original generated
file and removing it allows the file to open and be added properly, so this
is at least part of the issue. That "fix" worked on Android, but not iOS. I
am unsure of the function of the "UID:" line so I'm not sure what effect
its removal would have on overall functionality or what other issues may be
contributing.

**Comment by commonpike** [comment]
OK .. are you saying the ics download doesn't work on mobile, both android and ios ? and that that is because of the UID line ? We should create a new issue.

The UID doesn't do anything, but it was part of the spec - I am supposing it makes sure, if you add the event to your calendar a second time, it doesn't add it twice but recognises that it has already been added. Which is useful. Would be nice to know what part of the UID is failing on mobile.

**Comment by LittleDogFido** [comment]
So after some more troubleshooting it seems to problem was not what it
appeared and in fact there may not be any real issue at all. The act of
downloading the ics file worked fine of both Android and iOS.

When I was attempting to to open the file on Android I get an error message
which reads "Unable to launch this event". Opening up the ICS file in the
text editor it looks identical to the one downloaded on a desktop. This
"unable to launch" error is what is resolved by removing UID line, after
which it works fine. I had made sure before attempting to add the event
that it was not already on the calendar. HOWEVER, I found that Google
Calendar was the default calendar app on my phone and the downloading  and
adding of the event worked fine when the phone(Samsung)'s own calendar app
was used to open, regardless of the UID. So while the ICS file doesn't work
with the UID line on the Google Calendar, if someone were using to click on
the add-to-calendar drop-down on their phone they would likely be clicking
the "Google" link anyways which works fine with the Google Calendar app. In
regards to the Outlook app, after doing some research it seems that the
mobile Outlook app does not currently support ics imports, so again that is
an app issue and not something related to the ics file being generated.
https://outlook.uservoice.com/forums/293346-outlook-for-android/suggestions/9623127-support-for-ics-calendar-files

The troubleshooting I did on iOS again shows that there aren't really any
major issues with functionality. It downloads and adds to the calendar just
fine. The one exception to this is if the event had previously been added
and then deleted from the calendar. In these cases, the process of opening
the file and re-adding to the calendar appears goes smoothly, but it still
will not show up in the calendar. Removing the UID line does not resolve
this issue and doing so displays a message of "This invitation is out of
date" where the in-app button to add it to the calendar should be. So again
I'm inclined to think the problem is with the iOS calendar app and not the
file.

So while there are some minor limitations in regards to mobile, they are
circumstantial app-related issues that I don't think will have a major
impact on user experience. I think the inability to add import ics files
into the Outlook app is probably the biggest issue though, and it may be
worth revisiting the troubleshooting process if and when the feature is
added.
